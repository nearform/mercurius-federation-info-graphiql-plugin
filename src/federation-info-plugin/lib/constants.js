// enum TypeKind from GraphQL-JS repo: /src/type/introspection.ts
export const TYPE_KIND = {
  SCALAR: 'SCALAR',
  OBJECT: 'OBJECT',
  INTERFACE: 'INTERFACE',
  UNION: 'UNION',
  ENUM: 'ENUM',
  INPUT_OBJECT: 'INPUT_OBJECT',
  LIST: 'LIST',
  NON_NULL: 'NON_NULL'
}

// introspectionTypes from GraphQL-JS repo: /src/type/introspection.ts
export const INTROSPECTION_TYPES = [
  '__Schema',
  '__Directive',
  '__DirectiveLocation',
  '__Type',
  '__Field',
  '__InputValue',
  '__EnumValue',
  '__TypeKind'
]

// specifiedScalarTypes from GraphQL-JS repo: /src/type/scalars.ts
export const SCALAR_TYPES = ['String', 'Int', 'Float', 'Boolean', 'ID']

// Observed from examples
export const GRAPH_QL_QUERY_NAMES = ['_entities', '_service']
export const GRAPH_QL_OBJECT_NAMES = ['_Service']
