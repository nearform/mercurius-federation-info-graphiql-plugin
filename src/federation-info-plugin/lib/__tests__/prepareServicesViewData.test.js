import { IGNORED_FIELDS } from '../constants'
import { buildSchema, introspectionFromSchema } from 'graphql'
import {
  indexServiceFields,
  prepareServicesViewData
} from '../prepareServicesViewData'

const fieldsFromNames = names =>
  names.map(name => ({
    name,
    type: {
      name: 'String'
    }
  }))

describe('indexServiceFields', () => {
  it('returns indexed fields', () => {
    const fields = fieldsFromNames([...IGNORED_FIELDS, 'test'])

    expect(indexServiceFields(fields, 'dummyService')).toEqual({
      test: {
        name: 'test',
        type: { name: 'String' },
        serviceName: 'dummyService',
        typeString: 'String'
      }
    })
  })
})

describe('prepareServicesViewData', () => {
  it('Checks correct result', () => {
    const testService = introspectionFromSchema(
      buildSchema(
        `
          type Query {
            test(customInput: CustomInput): CustomType
          }

          type CustomType {
            id: ID
            name: String
          }

          input CustomInput {
            names: [String]!
          }
          
          type _Entity {
            name: String
          }

          enum TestEnum {
            one
            two
            three
          }
        `
      )
    )
    const servicesResult = { services: { testService } }
    expect(prepareServicesViewData(servicesResult)).toMatchSnapshot()

    // Potential replacements for snapshot test
    // Object.keys(itemsMap) equals ['Query', 'CustomType', 'CustomInput', 'TestEnum']

    // itemsMap.Query.itemsMap.test

    // itemsMap.CustomInput.itemsMap.names
    // itemsMap.CustomInput.itemsMap.names.typeString": "[String]!"
  })
})
