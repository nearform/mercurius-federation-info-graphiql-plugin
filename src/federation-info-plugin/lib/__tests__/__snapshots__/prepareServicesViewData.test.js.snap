// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prepareServicesViewData Checks correct result 1`] = `
[
  {
    "itemsMap": {
      "CustomInput": {
        "description": null,
        "enumValues": null,
        "fields": null,
        "inputFields": [
          {
            "defaultValue": null,
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "names",
            "serviceName": "testService",
            "type": {
              "kind": "NON_NULL",
              "name": null,
              "ofType": {
                "kind": "LIST",
                "name": null,
                "ofType": {
                  "kind": "SCALAR",
                  "name": "String",
                  "ofType": null,
                },
              },
            },
            "typeString": "[String]!",
          },
        ],
        "interfaces": null,
        "itemsMap": {
          "names": {
            "defaultValue": null,
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "names",
            "serviceName": "testService",
            "type": {
              "kind": "NON_NULL",
              "name": null,
              "ofType": {
                "kind": "LIST",
                "name": null,
                "ofType": {
                  "kind": "SCALAR",
                  "name": "String",
                  "ofType": null,
                },
              },
            },
            "typeString": "[String]!",
          },
        },
        "kind": "INPUT_OBJECT",
        "name": "CustomInput",
        "serviceName": "testService",
        "possibleTypes": null,
        "specifiedByURL": null,
      },
      "CustomType": {
        "description": null,
        "enumValues": null,
        "fields": [
          {
            "args": [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "id",
            "serviceName": "testService",
            "type": {
              "kind": "SCALAR",
              "name": "ID",
              "ofType": null,
            },
            "typeString": "ID",
          },
          {
            "args": [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "name",
            "serviceName": "testService",
            "type": {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
            "typeString": "String",
          },
        ],
        "inputFields": null,
        "interfaces": [],
        "itemsMap": {
          "id": {
            "args": [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "id",
            "serviceName": "testService",
            "type": {
              "kind": "SCALAR",
              "name": "ID",
              "ofType": null,
            },
            "typeString": "ID",
          },
          "name": {
            "args": [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "name",
            "serviceName": "testService",
            "type": {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
            "typeString": "String",
          },
        },
        "kind": "OBJECT",
        "name": "CustomType",
        "serviceName": "testService",
        "possibleTypes": null,
        "specifiedByURL": null,
      },
      "Query": {
        "description": null,
        "enumValues": null,
        "fields": [
          {
            "args": [
              {
                "defaultValue": null,
                "deprecationReason": null,
                "description": null,
                "isDeprecated": false,
                "name": "customInput",
                "type": {
                  "kind": "INPUT_OBJECT",
                  "name": "CustomInput",
                  "ofType": null,
                },
              },
            ],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "test",
            "serviceName": "testService",
            "type": {
              "kind": "OBJECT",
              "name": "CustomType",
              "ofType": null,
            },
            "typeString": "CustomType",
          },
        ],
        "inputFields": null,
        "interfaces": [],
        "itemsMap": {
          "test": {
            "args": [
              {
                "defaultValue": null,
                "deprecationReason": null,
                "description": null,
                "isDeprecated": false,
                "name": "customInput",
                "type": {
                  "kind": "INPUT_OBJECT",
                  "name": "CustomInput",
                  "ofType": null,
                },
              },
            ],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "test",
            "serviceName": "testService",
            "type": {
              "kind": "OBJECT",
              "name": "CustomType",
              "ofType": null,
            },
            "typeString": "CustomType",
          },
        },
        "kind": "OBJECT",
        "name": "Query",
        "serviceName": "testService",
        "possibleTypes": null,
        "specifiedByURL": null,
      },
      "TestEnum": {
        "description": null,
        "enumValues": [
          {
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "one",
          },
          {
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "two",
          },
          {
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "three",
          },
        ],
        "fields": null,
        "inputFields": null,
        "interfaces": null,
        "itemsMap": {},
        "kind": "ENUM",
        "name": "TestEnum",
        "serviceName": "testService",
        "possibleTypes": null,
        "specifiedByURL": null,
      },
    },
    "serviceName": "testService",
  },
]
`;

exports[`prepareServicesViewData Checks correct result 1`] = `
[
  {
    "itemsMap": {
      "CustomInput": {
        "description": null,
        "enumValues": null,
        "fields": null,
        "inputFields": [
          {
            "defaultValue": null,
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "names",
            "serviceName": "testService",
            "type": {
              "kind": "NON_NULL",
              "name": null,
              "ofType": {
                "kind": "LIST",
                "name": null,
                "ofType": {
                  "kind": "SCALAR",
                  "name": "String",
                  "ofType": null,
                },
              },
            },
            "typeString": "[String]!",
          },
        ],
        "interfaces": null,
        "itemsMap": {
          "names": {
            "defaultValue": null,
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "names",
            "serviceName": "testService",
            "type": {
              "kind": "NON_NULL",
              "name": null,
              "ofType": {
                "kind": "LIST",
                "name": null,
                "ofType": {
                  "kind": "SCALAR",
                  "name": "String",
                  "ofType": null,
                },
              },
            },
            "typeString": "[String]!",
          },
        },
        "kind": "INPUT_OBJECT",
        "name": "CustomInput",
        "possibleTypes": null,
        "serviceName": "testService",
        "specifiedByURL": null,
      },
      "CustomType": {
        "description": null,
        "enumValues": null,
        "fields": [
          {
            "args": [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "id",
            "serviceName": "testService",
            "type": {
              "kind": "SCALAR",
              "name": "ID",
              "ofType": null,
            },
            "typeString": "ID",
          },
          {
            "args": [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "name",
            "serviceName": "testService",
            "type": {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
            "typeString": "String",
          },
        ],
        "inputFields": null,
        "interfaces": [],
        "itemsMap": {
          "id": {
            "args": [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "id",
            "serviceName": "testService",
            "type": {
              "kind": "SCALAR",
              "name": "ID",
              "ofType": null,
            },
            "typeString": "ID",
          },
          "name": {
            "args": [],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "name",
            "serviceName": "testService",
            "type": {
              "kind": "SCALAR",
              "name": "String",
              "ofType": null,
            },
            "typeString": "String",
          },
        },
        "kind": "OBJECT",
        "name": "CustomType",
        "possibleTypes": null,
        "serviceName": "testService",
        "specifiedByURL": null,
      },
      "Query": {
        "description": null,
        "enumValues": null,
        "fields": [
          {
            "args": [
              {
                "defaultValue": null,
                "deprecationReason": null,
                "description": null,
                "isDeprecated": false,
                "name": "customInput",
                "type": {
                  "kind": "INPUT_OBJECT",
                  "name": "CustomInput",
                  "ofType": null,
                },
              },
            ],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "test",
            "serviceName": "testService",
            "type": {
              "kind": "OBJECT",
              "name": "CustomType",
              "ofType": null,
            },
            "typeString": "CustomType",
          },
        ],
        "inputFields": null,
        "interfaces": [],
        "itemsMap": {
          "test": {
            "args": [
              {
                "defaultValue": null,
                "deprecationReason": null,
                "description": null,
                "isDeprecated": false,
                "name": "customInput",
                "type": {
                  "kind": "INPUT_OBJECT",
                  "name": "CustomInput",
                  "ofType": null,
                },
              },
            ],
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "test",
            "serviceName": "testService",
            "type": {
              "kind": "OBJECT",
              "name": "CustomType",
              "ofType": null,
            },
            "typeString": "CustomType",
          },
        },
        "kind": "OBJECT",
        "name": "Query",
        "possibleTypes": null,
        "serviceName": "testService",
        "specifiedByURL": null,
      },
      "TestEnum": {
        "description": null,
        "enumValues": [
          {
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "one",
          },
          {
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "two",
          },
          {
            "deprecationReason": null,
            "description": null,
            "isDeprecated": false,
            "name": "three",
          },
        ],
        "fields": null,
        "inputFields": null,
        "interfaces": null,
        "itemsMap": {},
        "kind": "ENUM",
        "name": "TestEnum",
        "possibleTypes": null,
        "serviceName": "testService",
        "specifiedByURL": null,
      },
    },
    "serviceName": "testService",
  },
]
`;
